### PR TITLE
Fix plugin deploy to respect return code from invoked bash scripts

### PR DESCRIPTION
**Pull request by PhillypHenning**
# Description
Updates the run_cmd function in `utilities.py` to respect the invoked bash script return code

## Type of change

Please delete options that are not relevant.

- [X] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?
This was tested locally by building a local bitops and mounting my local terraform plugin with the update to the plugin deploy.sh that makes it `exit 1`

**Logs**
```
+ exit 1
2022-12-07 17:27:54,304 root         ERROR <subprocess.Popen object at 0x7f59e1395be0>
```

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation

### CODE REVIEW

**Comment by PhillypHenning** [comment]
fixes https://github.com/bitovi/bitops/issues/358

**Review by armab — CHANGES_REQUESTED** [review]
Considering we can only report error like this in the function itself, it's not helpful:
> 2022-12-07 17:27:54,304 root ERROR <subprocess.Popen object at 0x7f59e1395be0>

The error handling and reporting should be done on another layer, - code that's using this function. The function `run_cmd` returns a `subprocess.CompletedProcess` instance to be handled in next steps.

Here is the spot: 
https://github.com/bitovi/bitops/blob/5b0fec06e6f74c9aa5d1e63509215859ec230f82/scripts/plugins/deploy_plugins.py#L271-L281

File: scripts/plugins/utilities.py [hunk 1]
```diff
@@ -369,6 +369,11 @@ def run_cmd(command: Union[list, str]) -> subprocess.CompletedProcess:
             # about the status of the process execution.
             # Namely the return code which is used elsewhere.
             process.communicate()
+            process_return_code = process.returncode
+            if process_return_code != 0:
+                logger.error(process)
```
[the remark this answers is not in the document]
Comment (PhillypHenning) [reply scripts/plugins/utilities.py:374]: It did print out the error above it. 
```
+ exit 1
```

I purposely stopped the process with `exit 1`
I may be misunderstanding but I'm not seeing why this should be moved. 

That being said, thank you looking at it and bringing it up as it should also consider fail_fast

[the remark this answers is not in the document]
Comment (armab) [reply scripts/plugins/utilities.py:374]: the `process` here is of `subprocess.CompletedProcess` data type.

Do you have an example of an Error log when this gets printed in the output?
```
logger.error(process)
```
How does the error might look in the real world when we log it here ^^?

[the remark this answers is not in the document]
Comment (armab) [reply scripts/plugins/utilities.py:374]: The `run_cmd` is an utility function that returns `subprocess.CompletedProcess` instance.
It's used in 2 places:
1) 
https://github.com/bitovi/bitops/blob/5b0fec06e6f74c9aa5d1e63509215859ec230f82/scripts/plugins/install_plugins.py#L188-L201

2) 
https://github.com/bitovi/bitops/blob/5b0fec06e6f74c9aa5d1e63509215859ec230f82/scripts/plugins/deploy_plugins.py#L271-L281



In the code snippets above, 1st example works well, 2nd example has a bug. That's the right place for a fix, not in the function. We want to handle the exit code outside of the function to have enough context for useful logging error. If we exit the function early, we never reach the code which prints these (user-friendly) errors: 

https://github.com/bitovi/bitops/blob/5b0fec06e6f74c9aa5d1e63509215859ec230f82/scripts/plugins/install_plugins.py#L196
https://github.com/bitovi/bitops/blob/5b0fec06e6f74c9aa5d1e63509215859ec230f82/scripts/plugins/deploy_plugins.py#L281

[the remark this answers is not in the document]
Comment (PhillypHenning) [reply scripts/plugins/utilities.py:374]: You're incorrect. 
The `communicate` function handles the interaction of invoking. 
The communicate builds the CompletedProcess Object, and you return the popen process, which contains the CompletedProcess Object as an attribute. 


```
Popen.communicate(input=None, timeout=None)[¶](https://docs.python.org/3/library/subprocess.html#subprocess.Popen.communicate)

    Interact with process: Send data to stdin. Read data from stdout and stderr, until end-of-file is reached. Wait for process to terminate and set the [returncode](https://docs.python.org/3/library/subprocess.html#subprocess.Popen.returncode) attribute.
    ```

You're passing back the results of Popen through the return, you do not pass back the CompletedProcess object, that is an attribute within the Popen object. 

I stand by the placement, I do not believe it should be moved.

[the remark this answers is not in the document]
Comment (PhillypHenning) [reply scripts/plugins/utilities.py:374]: Though I do understand what you're saying above. 

What you're asking for is; 
Raise an exception in `run_cmd` based on the return code. 

That would mean that anything that calls `run_cmd` would need to consider how to handle errors around it. 

I'm not exactly bothered by this, but I fail to see what value that adds is all

File: scripts/plugins/utilities.py [hunk 2]
```diff
@@ -369,9 +372,8 @@ def run_cmd(command: Union[list, str]) -> subprocess.CompletedProcess:
...
             # Namely the return code which is used elsewhere.
             process.communicate()
+
     except Exception as exc:
         logger.error(exc)
-        if BITOPS_fast_fail_mode:
-            sys.exit(101)
-
+        sys.exit(101)
     return process
```
Comment (armab) [inline scripts/plugins/utilities.py:379]: The only thing we need to ensure the correct data type is specified here: 

https://github.com/bitovi/bitops/blob/dab9410f793a56a37105f6232e11c6616432a3a8/scripts/plugins/utilities.py#L352-L353

I think it's `subprocess.Popen` now.

Comment (PhillypHenning) [reply scripts/plugins/utilities.py:379]: **Log**
`2022-12-09 14:13:05,519 root         CRITICAL <class 'subprocess.Popen'>`

**Tested with**
```
process.communicate()
logger.critical(type(process))
```

I agree with the change 👍

File: scripts/plugins/utilities.py [hunk 3]
```diff
@@ -348,10 +348,13 @@ def handle_hooks(mode, hooks_folder, source_folder):
...
             logger.debug(result.stdout)
+            if BITOPS_fast_fail_mode:
+                sys.exit(result.returncode)
+
     os.chdir(original_directory)
 
 
-def run_cmd(command: Union[list, str]) -> subprocess.CompletedProcess:
+def run_cmd(command: Union[list, str]) -> subprocess.Popen:
     """Run a linux command and return CompletedProcess instance as a result"""
```
Comment (armab) [inline scripts/plugins/utilities.py:358]: Also missing a correct `Popen` reference in the function description: 
```suggestion
    """Run a linux command and return subprocess.Popen instance as a result"""
```

outside of that, looks good!

Comment (PhillypHenning) [reply scripts/plugins/utilities.py:358]: Shoot sorry

**Review by armab — APPROVED** [review]
Looks good, thanks for the bug fixes 👍

**Comment by PhillypHenning** [comment]
Reason for auto-merge cancel
The effort of reworking install_plugins is way beyond this tickets scope.

**Pull request merged**